src/gensxg.c

```diff
@@ -0,0 +1,377 @@
+    EVP_PKEY* pub_key = load_ed25519_pubkey(opt->public_key);
+    EVP_PKEY* pri_key = load_private_key(opt->private_key);
+    if (!sxg_add_ed25519_signer(opt->url, opt->date, opt->expire,
+                                opt->validity_url, pri_key,
+                                pub_key, signers)) {
+      fprintf(stderr, "Failed to generate SXG signer with Ed25519 key\n");
+      exit(1);
+    }
+    EVP_PKEY_free(pub_key);
+    EVP_PKEY_free(pri_key);
+  }
+}
+
+static void load_contents(const Options* opt, sxg_buffer_t* buf) {
+  FILE* file = fopen(opt->content, "rb");
+  if (file == NULL) {
+    fprintf(stderr, "Failed to open content: %s\n", opt->content);
+    exit(1);
+  }
+  if (fseek(file, 0, SEEK_END) != 0) {
+    fprintf(stderr, "Failed to seek content: %s\n", opt->content);
+    exit(1);
+  }
+  long filesize = ftell(file);
+  if (filesize == -1) {
+    fprintf(stderr, "Failed to load content: %s\n", opt->content);
+    exit(1);
+  }
+  rewind(file);
+  sxg_buffer_resize(filesize, buf);
+  fread(buf->data, sizeof(uint8_t), filesize, file);
+  fclose(file);
+}
+
+static void print_help() {
+  fprintf(stderr, kHelpMessage);
+}
+
+static void dump(const Options* opt) {
+  fprintf(stderr,
+          "Input arguments: \n"
+          " url: %s\n"
+          " certUrl: %s\n"
+          " validityUrl: %s\n"
+          " certificate: %s\n"
+          " publicKey: %s\n"
+          " privateKey: %s\n"
+          " content: %s\n"
+          " contentType: %s\n"
+          " date: %ld\n"
+          " expire: %ld\n"
+          " miRecordSize: %ld\n",
+          opt->url,
+          opt->cert_url,
+          opt->validity_url,
+          opt->certificate,
+          opt->public_key,
+          opt->private_key,
+          opt->content,
+          opt->content_type,
+          opt->date,
+          opt->expire,
+          opt->mi_record_size);
+}
+
+int main(int argc, char* const argv[]) {
+  Options opt = parse_options(argc, argv);
+  if (opt.help) {
+    print_help();
+    return 0;
+  }
+
+  if (!validate(&opt)) {
+    dump(&opt);
+    return 1;
+  }
+  
+  sxg_signer_list_t signers = sxg_empty_signer_list();;
+  load_signer(&opt, &signers);
+  
+  sxg_raw_response_t content = sxg_empty_raw_response();;
+  load_contents(&opt, &content.payload);
+  if (!sxg_header_append_string("content-type", opt.content_type,
+                                &content.header)) {
+    fprintf(stderr, "Failed to append content-type header.\n");
+    return 1;
+  }
+  
+  sxg_encoded_response_t encoded = sxg_empty_encoded_response();
+  if (!sxg_encode_response(opt.mi_record_size, &content, &encoded)) {
+    fprintf(stderr, "Failed to encode contents.\n");
+    return 1;
+  }
+  
+  sxg_buffer_t result = sxg_empty_buffer();
+  if (!sxg_generate(opt.url, &signers, &encoded, &result)) {
+    fprintf(stderr, "Failed to generate SXG.\n");
+    return 1;
+  }
+
+  if (freopen(NULL, "wb", stdout) == NULL) {
+    fprintf(stderr, "Failed to reopen stdout.\n");
```

Comment:
[perror](https://en.cppreference.com/w/c/io/perror) could be a better choice. Maybe the same for some other errors.

Reply:
OK, using perror and strerror.